### PR TITLE
Remove handles from ves_icall_System_Array_FastCopy.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
The goal is reduction/elimination of both kinds of coop handles.
Eliminating the native ones is more interesting.
Eliminating the managed ones is easier.

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

File: mono/metadata/icall-decl.h [hunk 1]
```diff
@@ -166,9 +166,15 @@ ICALL_EXPORT int ves_icall_System_Threading_Thread_SystemMaxStackSize (void);
...
-ICALL_EXPORT void ves_icall_System_Buffer_MemcpyInternal (gpointer dest, gconstpointer src, gint32 count);
+
+ICALL_EXPORT void	 ves_icall_System_Array_GetGenericValue_icall (MonoArray**, guint32, gpointer);
+ICALL_EXPORT void	 ves_icall_System_Array_SetGenericValue_icall (MonoArray**, guint32, gpointer);
+ICALL_EXPORT void	 ves_icall_System_Array_InternalCreate (MonoArray *volatile* result, MonoType* type, gint32 rank, gint32* pLengths, gint32* pLowerBounds);
+ICALL_EXPORT MonoBoolean ves_icall_System_Array_CanChangePrimitive (MonoReflectionType *volatile* ref_src_type_handle, MonoReflectionType *volatile* ref_dst_type_handle, MonoBoolean reliable);
+ICALL_EXPORT MonoBoolean ves_icall_System_Array_FastCopy (MonoArray*volatile* source_handle, int source_idx, MonoArray*volatile* dest_handle, int dest_idx, int length);
+
+ICALL_EXPORT void	 ves_icall_System_Buffer_MemcpyInternal (gpointer dest, gconstpointer src, gint32 count);
+
```
Comment (lambdageek) [inline mono/metadata/icall-decl.h]: Unrelated whitespace changes.

The only actual change is adding a declaration for `ves_icall_System_Array_FastCopy` everything else is noise.

File: mono/metadata/icall-decl.h [hunk 2]
```diff
@@ -221,8 +227,7 @@ ICALL_EXPORT void ves_icall_Mono_Runtime_AnnotateMicrosoftTelemetry (const char
 ICALL_EXPORT void ves_icall_Mono_Runtime_DisableMicrosoftTelemetry (void);
 ICALL_EXPORT int ves_icall_Mono_Runtime_CheckCrashReportingLog (const char *directory, MonoBoolean clear);
 ICALL_EXPORT void ves_icall_Mono_Runtime_EnableCrashReportingLog (const char *directory);
-ICALL_EXPORT void ves_icall_System_Array_InternalCreate (MonoArray *volatile* result, MonoType* type, gint32 rank, gint32* pLengths, gint32* pLowerBounds);
-ICALL_EXPORT MonoBoolean ves_icall_System_Array_CanChangePrimitive (MonoReflectionType *volatile* ref_src_type_handle, MonoReflectionType *volatile* ref_dst_type_handle, MonoBoolean reliable);
+
```
Comment (lambdageek) [inline mono/metadata/icall-decl.h]: Unrelated line damage

File: mcs/class/corlib/System/Array.cs [hunk 3]
```diff
@@ -583,7 +583,7 @@ public static void Copy (Array sourceArray, int sourceIndex, Array destinationAr
 			if (destinationIndex < 0)
 				throw new ArgumentOutOfRangeException ("destinationIndex", "Value has to be >= 0.");
 
-			if (FastCopy (sourceArray, sourceIndex, destinationArray, destinationIndex, length))
+			if (FastCopy (ref sourceArray, sourceIndex, ref destinationArray, destinationIndex, length))
```
Comment (lambdageek) [inline mcs/class/corlib/System/Array.cs]: I wonder if we need extra locals here.

If we had a caller like `Copy (o.member1, 0, o.member2, 0, len)` and then `Copy` is inlined, won't we potentially end up with passing a heap address to `FastCopy` here? 

Maybe we should have `FastCopy` that assigns the params to locals and have it call a private `FastCopy_icall` with ref arguments.

File: mono/metadata/icall-decl.h (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/icall-decl.h]: This file is constantly getting merge conflicts. A little organization instead of always appending will help reduce that.

File: mcs/class/corlib/System/Array.cs (hunk 3 above)
Comment (jaykrell) [reply mcs/class/corlib/System/Array.cs]: Oh good point. Mark it as noinline?

Comment (jaykrell) [reply mcs/class/corlib/System/Array.cs]: But does it matter? If only read? Maybe start adding const to “ref ref”?

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
I did it your way.

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

File: mcs/class/corlib/System/Array.cs [hunk 4]
```diff
@@ -364,7 +364,13 @@ void IEnumerator.Reset ()
 		internal extern void SetValueImpl (object value, int pos);
 
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		internal extern static bool FastCopy (Array source, int source_idx, Array dest, int dest_idx, int length);
+		extern static bool FastCopy_icall (Array source, int source_idx, Array dest, int dest_idx, int length);
```
Comment (jaykrell) [inline mcs/class/corlib/System/Array.cs]: Darn oops needs ref.

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
The extra `_internal` vs. `_icall` layer is nice when there are "many" callers, less churn.
But also less efficient, alas.

File: mcs/class/corlib/System/Array.cs (hunk 3 above)
Comment (lambdageek) [reply mcs/class/corlib/System/Array.cs]: suppose we have
```
  class C {
     object member;
   }
  C c = new C ();
  c.member = new object ();
```

if we have `ref c.member` that's a pointer into the middle of `c`.  So we'll pin `c`.  But nothing is pinning the object that `c.member` currently points to. So the GC could move it.  so if on the native side you have `MonoObject * volatile * ptr` and if you're not careful about always using `*ptr` (instead of a local initialized with the present value of `*ptr`) the GC could move the object and you will be pointing at garbage.

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by akoeplinger** [comment]
The mono/mono repository is focused on stability and limiting code churn now so I don't think this change is worth doing here. If you think the change is important please reopen it in dotnet/runtime, thank you!
